lib/OpenQA/Schema/Result/Assets.pm

```diff
@@ -242,52 +243,108 @@ sub limit_assets {
+        my $assets = $asset_resultset->search(
             {
                 id => {-in => [map { $_->asset_id } @job_assets]}});
+
+        # assign assets to a hash
+        my %assets;
         while (my $a = $assets->next) {
             $assets{$a->id} = $a;
         }
+
+        # define variable to keep track of all the jobs which have been kept
+        my @kept_by_jobgroup;
+
+        # loop through all entries in the job-to-asset mapping
         for my $a (@job_assets) {
             my $asset = $assets{$a->asset_id};
+
             OpenQA::Utils::log_debug(
                 sprintf "Group %s: %s/%s %s->%s",
                 $g->name, $asset->type, $asset->name,
                 human_readable_size($asset->size // 0),
                 human_readable_size($sizelimit));
+
             # ignore fixed assets
             next if ($asset->is_fixed);
+
+            # add asset to mapping of seen assets by group
             $seen_asset{$asset->id} = $g->id;
+
+            # check whether the asset has a size and whether we haven't exceeded the reduce limit
             my $size = $asset->ensure_size;
             if ($size > 0 && $reduceto > 0) {
-                $keep{$a->asset_id} = sprintf "%s: %s/%s", $g->name, $asset->type, $asset->name;
+                # keep asset
+                push(
+                    @kept_by_jobgroup,
+                    $keep{$a->asset_id} = {
+                        group_id => $g->id,
+                        job_id   => $a->job_id,
+                        log_msg  => sprintf "%s: %s/%s",
+                        $g->name, $asset->type, $asset->name,
+                    });
             }
             else {
+                # add assets to removal list
                 $toremove{$a->asset_id} = sprintf "%s/%s", $asset->type, $asset->name;
+                # set flag for removal if we have exceeded the actual size limit
                 $doremove = 1 if ($sizelimit <= 0);
             }
+
             $sizelimit -= $size;
             $reduceto  -= $size;
         }
+
+        # print messages
+        if ($debug_keep && @kept_by_jobgroup) {
+            OpenQA::Utils::log_debug('the following assets are kept by job group: ' . $g->name);
+            for my $kept_asset (@kept_by_jobgroup) {
+                OpenQA::Utils::log_debug($kept_asset->{log_msg} . '; referencing job: ' . $kept_asset->{job_id});
+            }
+        }
     }
+
+    # remove all kept assets from the removal list and propagate last update to asset table
     for my $id (keys %keep) {
-        OpenQA::Utils::log_debug("KEEP $toremove{$id} $keep{$id}") if $debug_keep;
+        my $asset = $keep{$id};
+
+        OpenQA::Utils::log_debug("KEEP $toremove{$id} $asset->{log_msg}") if $debug_keep;
         delete $toremove{$id};
+
+        # set the time of the last use and the related job group
+        my $job = $jobs->find($asset->{job_id});
+        $asset_resultset->find($id)->update(
```

Comment:
I assume this find($id)->update happens in one SQL?  and why do you query $job here?

Reply:
In the first place I wanted to store the date of the last use directly in the asset table. This is a leftover, no reason for the query anymore.